fitbenchmarking/parsing/nist_data_functions.py

```diff
@@ -58,3 +60,108 @@ def format_function_scipy(function):
     function = function.replace("pi", "np.pi")
 
     return function
+
+
+def is_safe(func_str):
+    """
+    Verifies that a string is safe to be passed to exec in the context of an
+    equation.
+
+    :param func_str: The function to be checked
+    :type func_str: string
+
+    :returns: Whether the string is of the expected format for an equation
+    :rtype: bool
+    """
+    # Remove whitespace
+    while ' ' in func_str:
+        func_str = func_str.replace(' ', '')
+
+    # Empty string is safe
+    if func_str == '':
+        return True
+
+    # These are all safe and can be stripped out
+    if 'np' in func_str:
+        np_funcs = ['np.exp', 'np.cos', 'np.sin', 'np.tan', 'np.pi']
+        for s in np_funcs:
+            func_str = func_str.replace(s, '')
+
+    # Store valid symbols for later
+    symbols = ['**', '/', '*', '+', '-']
+
+    # Partition on outer brackets
+    if '(' in func_str:
+        if ')' not in func_str:
+            # Number of brackets don't match
```

Comment:
They check that there is a closing bracket for the opening bracket and since it recurses, it will be equivalent?
The only difference is in a case where the number is balanced, this will only have to search the string for one each time whereas your suggestion counts them all, and in the case where it's not balanced, your suggestion wins by stopping in the first iteration.

Reply:
Yeah fair enough, it's fine as it is.  At this point I hadn't noticed the recursion - I should have gone back and removed this comment.